connect-inject/endpoints_controller_test.go

```diff
@@ -0,0 +1,1322 @@
+			ep := &EndpointsController{
+				Client:                client,
+				Log:                   logrtest.TestLogger{T: t},
+				ConsulClient:          consulClient,
+				ConsulPort:            consulPort,
+				ConsulScheme:          "http",
+				AllowK8sNamespacesSet: mapset.NewSetWith("*"),
+				DenyK8sNamespacesSet:  mapset.NewSetWith(),
+				ReleaseName:           "consul",
+				ReleaseNamespace:      "default",
+			}
+			namespacedName := types.NamespacedName{
+				Namespace: "default",
+				Name:      "service-created",
+			}
+
+			resp, err := ep.Reconcile(ctrl.Request{
+				NamespacedName: namespacedName,
+			})
+			require.NoError(t, err)
+			require.False(t, resp.Requeue)
+
+			// After reconciliation, Consul should have the service with the correct number of instances
+			serviceInstances, _, err := consulClient.Catalog().Service(tt.consulSvcName, "", nil)
+			require.NoError(t, err)
+			require.Len(t, serviceInstances, tt.expectedNumSvcInstances)
+			for i, instance := range serviceInstances {
+				require.Equal(t, tt.expectedConsulSvcInstances[i].ServiceID, instance.ServiceID)
+				require.Equal(t, tt.expectedConsulSvcInstances[i].ServiceName, instance.ServiceName)
+				require.Equal(t, tt.expectedConsulSvcInstances[i].ServiceAddress, instance.ServiceAddress)
+				require.Equal(t, tt.expectedConsulSvcInstances[i].ServicePort, instance.ServicePort)
+				require.Equal(t, tt.expectedConsulSvcInstances[i].ServiceMeta, instance.ServiceMeta)
+				require.Equal(t, tt.expectedConsulSvcInstances[i].ServiceTags, instance.ServiceTags)
+			}
+			proxyServiceInstances, _, err := consulClient.Catalog().Service(fmt.Sprintf("%s-sidecar-proxy", tt.consulSvcName), "", nil)
+			require.NoError(t, err)
+			require.Len(t, proxyServiceInstances, tt.expectedNumSvcInstances)
+			for i, instance := range proxyServiceInstances {
+				require.Equal(t, tt.expectedProxySvcInstances[i].ServiceID, instance.ServiceID)
+				require.Equal(t, tt.expectedProxySvcInstances[i].ServiceName, instance.ServiceName)
+				require.Equal(t, tt.expectedProxySvcInstances[i].ServiceAddress, instance.ServiceAddress)
+				require.Equal(t, tt.expectedProxySvcInstances[i].ServicePort, instance.ServicePort)
+				require.Equal(t, tt.expectedProxySvcInstances[i].ServiceProxy, instance.ServiceProxy)
+				require.Equal(t, tt.expectedProxySvcInstances[i].ServiceMeta, instance.ServiceMeta)
+				require.Equal(t, tt.expectedProxySvcInstances[i].ServiceTags, instance.ServiceTags)
+			}
+
+			_, checkInfos, err := consulClient.Agent().AgentHealthServiceByName(fmt.Sprintf("%s-sidecar-proxy", tt.consulSvcName))
+			expectedChecks := []string{"Proxy Public Listener", "Destination Alias"}
+			require.NoError(t, err)
+			require.Len(t, checkInfos, tt.expectedNumSvcInstances)
+			for _, checkInfo := range checkInfos {
+				checks := checkInfo.Checks
+				require.Contains(t, expectedChecks, checks[0].Name)
+				require.Contains(t, expectedChecks, checks[1].Name)
+			}
```

Comment:
I think for completeness, we need an assertion for upstreams too

Reply:
The proxy service has the upstreams in it, and for the proxy service I am already asserting
`require.Equal(t, tt.expectedProxySvcInstances[i].ServiceProxy, instance.ServiceProxy)`.

If I change `expectedProxySvcInstances[0].ServiceProxy.Upstreams` (on L505) to have a different port, it does fail, so I think the test covers the upstreams annotation on L458. 

Let me know if you meant something different though!